Allow prune mode nodes to rewind past pruning horizon

SWvheerden opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
We need pruned nodes to be able to handle a reorg past their pruning horizon. This pr fixes prune mode to allow to happen as well as add a cucumber test proving this is possible. 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
This was tested as follows:
* Cucumber test. 
* Synced a node up to tip height of 19616 with a pruning horizon of 1000

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [x] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I ran `cargo test` successfully before submitting my PR.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

stringhandler commented on base_layer/core/src/chain_storage/async_db.rs:
```diff
@@ -313,8 +313,8 @@ impl<'a, B: BlockchainBackend + 'static> AsyncDbTransaction<'a, B> {
 
-    pub fn update_kernel_sum(&mut self, header_hash: HashOutput, kernel_sum: Commitment) -> &mut Self {
-        self.transaction.update_kernel_sum(header_hash, kernel_sum);
+    pub fn update_deleted_with_diff(&mut self, header_hash: HashOutput, deleted: Bitmap) -> &mut Self {
```
Rather rename the `update_deteled` method

stringhandler commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -1471,11 +1472,15 @@ fn rewind_to_height<T: BlockchainBackend>(db: &mut T, height: u64) -> Result<Vec
+    let (last_header, header_accumulated_data) = db.fetch_header_and_accumulated_data(height)?;
+
+    if steps_back > metadata.pruning_horizon() && metadata.pruning_horizon() > 0 {
+        debug!(target: LOG_TARGET, "WARNING, reorg past pruning horizon");
```
Why not make this a `warn`

stringhandler commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -332,15 +332,40 @@ impl LMDBDatabase {
                     block_accum_data.deleted = DeletedBitmap { deleted };
                     self.update_block_accumulated_data(&write_txn, height, &block_accum_data)?;
                 },
+                UpdateDeletedBlockAccumulatedDataWithDiff {
```
The `UpdateDeletedBlockAccumulateData` enum variant is most likely not used anymore and should be deleted

stringhandler commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -332,15 +332,40 @@ impl LMDBDatabase {
+                        self.fetch_block_accumulated_data(&write_txn, height - 1)?
+                            .unwrap_or_else(BlockAccumulatedData::default)
+                    } else {
+                        BlockAccumulatedData::default()
```
nit: Probably better to throw an Error (Attempt to update genesis block)

stringhandler reviewed (commented):
Nice so far, but I don't think this entirely solves the problem. 

The cucumber tests a reorg, but you really should be testing this situation:
1. The reorg happens at point R on ChainR, with chains ChainRA and ChainRB
2. On ChainRA, there is a spend of an output in ChainR, but in ChainRB that output is not spent.
3. When reorging to ChainRB, the output is no longer spent.

I am not sure that your PR solves this adequately, but I could be wrong. As far as I see it, the output will be pruned in ChainRB and unrecoverable

stringhandler commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -1472,23 +1475,55 @@ fn rewind_to_height<T: BlockchainBackend>(db: &mut T, height: u64) -> Result<Vec
+    let mut prune_past_horizon = false;
     if steps_back > metadata.pruning_horizon() && metadata.pruning_horizon() > 0 {
-        debug!(target: LOG_TARGET, "WARNING, reorg past pruning horizon");
+        info!(target: LOG_TARGET, "WARNING, reorg past pruning horizon");
```
Why not make this a `warn!`

stringhandler commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -1472,23 +1475,55 @@ fn rewind_to_height<T: BlockchainBackend>(db: &mut T, height: u64) -> Result<Vec
     if steps_back > metadata.pruning_horizon() && metadata.pruning_horizon() > 0 {
-        debug!(target: LOG_TARGET, "WARNING, reorg past pruning horizon");
+        info!(target: LOG_TARGET, "WARNING, reorg past pruning horizon");
+        info!(target: LOG_TARGET, "WARNING, rewinding back to 0");
```
suggestion: Combine this with the previous line

stringhandler commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -1471,20 +1475,55 @@ fn rewind_to_height<T: BlockchainBackend>(db: &mut T, height: u64) -> Result<Vec
         txn.delete_block(block.block.hash());
         txn.delete_header(last_block_height - h);
-        txn.insert_chained_orphan(block.clone());
+        if !prune_past_horizon {
```
None of these blocks can be pruned, otherwise they would error at `try_into_chain_block()`, so no need for this check

SWvheerden replied:
Yes, these blocks can't be pruned but it does not help if save block 8,9,10 in the orphan DB, but we delete 5,6,7 for example. 

This is here if we know we are rewinding behind the pruning horizon. We know we are destroying our "big block" so we need to recreate it anyway. So no need in saving these.  Archival nodes save these to maybe swing back to them if the chain fails syncing or the like.

stringhandler commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -1471,20 +1475,55 @@ fn rewind_to_height<T: BlockchainBackend>(db: &mut T, height: u64) -> Result<Vec
+            let block = fetch_block(db, last_block_height - h - steps_back)?;
+            let block = Arc::new(block.clone().into_pruned_chain_block());
+            txn.delete_block(block.block.hash());
+            removed_blocks.push(block);
```
This is potentially a LOT of blocks in memory

SWvheerden replied:
Yip, I am in two minds on this, keep it like this, or just don't give these to the mempool. Or we trigger a mempool wipe or the like.

stringhandler commented on integration_tests/features/Reorgs.feature:
```diff
@@ -30,5 +30,62 @@ Feature: Reorgs
+    Given I have a base node NODE1 connected to all seed nodes    
+    When I mine a block on NODE1 with coinbase CB1
+    Given I have a base node NODE2 connected to node NODE1
+    When I mine but don't submit a block BLOCK2 on NODE2
```
nit: this can be rolled into one step with some clever javascript

stringhandler commented on integration_tests/features/Reorgs.feature:
```diff
@@ -30,5 +30,62 @@ Feature: Reorgs
+    Given I have a pruned node PNODE2 connected to node NODE1
+    Then node PNODE2 is at height 10
+    When I start NODE2
+    When I submit block BLOCK2 to NODE2
```
same here

stringhandler replied:
When I submit blocks BLOCK2,BLOCK3,BLOCK4 to NODE2

stringhandler commented on integration_tests/features/Reorgs.feature:
```diff
@@ -30,5 +30,62 @@ Feature: Reorgs
+    When I submit transaction TX1 to NODE1
+    Then NODE1 has TX1 in MEMPOOL state
+    When I mine 6 blocks on NODE1
+    Given I have a pruned node PNODE2 connected to node NODE1
```
Might be more clear to explicitly specify what the pruning horizon is here so that it can be verified without opening the js code

stringhandler replied:
it also prevents it from breaking if the default is changed for another test

stringhandler commented on integration_tests/features/Reorgs.feature:
```diff
@@ -30,5 +30,62 @@ Feature: Reorgs
+    When I submit block BLOCK18 to NODE2
+    When I submit block BLOCK19 to NODE2
+    When I submit block BLOCK20 to NODE2
+    Then node NODE2 is at height 20
```
Then all nodes are at height 20

stringhandler commented on integration_tests/features/Reorgs.feature:
```diff
@@ -30,5 +30,62 @@ Feature: Reorgs
+    When I mine but don't submit a block BLOCK15 on NODE2
+    When I mine but don't submit a block BLOCK16 on NODE2
+    When I mine but don't submit a block BLOCK17 on NODE2
+    When I mine but don't submit a block BLOCK18 on NODE2
```
Aren't all of these blocks mined at height 2? Did this test pass?

SWvheerden replied:
No, what happens here, is NODE2 accepts these blocks and it moves forward in the chain.  This just stops the block from propagating.

stringhandler commented on integration_tests/features/Reorgs.feature:
```diff
@@ -30,5 +30,62 @@ Feature: Reorgs
+    When I mine but don't submit a block BLOCK19 on NODE2
+    When I mine but don't submit a block BLOCK20 on NODE2   
+    And I stop NODE2
+    When I mine 3 blocks on NODE1
```
I don't know if this is necessary

SWvheerden replied:
It's here to ensure `CB1` is past lock height

stringhandler commented on integration_tests/features/Reorgs.feature:
```diff
@@ -30,5 +30,62 @@ Feature: Reorgs
+    Then node PNODE2 is at height 20
+    # Because TX1 should have been re_orged out we should be able to spend CB1 again
+    When I create a transaction TX2 spending CB1 to UTX2
+    When I submit transaction TX2 to NODE1
```
```suggestion
    When I submit transaction TX2 to PNODE1
```

stringhandler commented on integration_tests/features/Reorgs.feature:
```diff
@@ -30,5 +30,62 @@ Feature: Reorgs
+    # Because TX1 should have been re_orged out we should be able to spend CB1 again
+    When I create a transaction TX2 spending CB1 to UTX2
+    When I submit transaction TX2 to NODE1
+    Then NODE1 has TX2 in MEMPOOL state
```
NODE1 is the wrong node to test here, you should be testing that the pruned node accepts it

SWvheerden replied:
NODE1 and PNODE1 should have the same chain, butt swapping it over.

stringhandler commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -1382,14 +1393,25 @@ impl BlockchainBackend for LMDBDatabase {
+                // integrity check, the diff should always be smaller than the total
+                let total_size = block_accum_data.cardinality();
+                let diff_bitmap = block_accum_data.xor(&prev_block_accum_data);
+                assert!(diff_bitmap.cardinality() <= total_size);
```
This is not the place for an assert
```suggestion
               
```

stringhandler reviewed (dismissed):
Happy, will test locally

SWvheerden committed:
Rewind prune mode behind horizon

Cleanup comments and prune mode
Fix error message in rpc
Send  bitmap diff and not the entire bitmap
fix rpc message loggging
update deleted with diff

Commit:
Merge branch 'development' into sw_re-org_behind_prune_hor
